File: tags.py
# -*- coding: utf-8 -*-

# Link a file into a tags directory
import os

class Tags(object):
    """
    Tag files to a folder
    """
    def __init__(s, root):
        s.root = root

    """
    Link a file
    """
    def link(s, filepath, tag):
        if os.path.isfile(filepath):
            filename = os.path.basename(filepath)
            path = os.path.join(s.root, tag)
            if not os.path.isdir(path):
                os.makedirs(path)
            path = os.path.join(path, filename)
            if not os.path.isfile(path):
                os.link(filepath, path)
            return path
        return None


# Command Line functionality
if __name__ == "__main__":
    import os
    from argparse import ArgumentParser, FileType
    parser = ArgumentParser(description="Create a hardlink to a file")
    parser.add_argument("file", help="The file you wish to extract from", type=str)
    parser.add_argument("tag", help="Tag name", type=str)
    args = parser.parse_args()
    root = os.getcwd()
    path = os.path.realpath(os.path.join(root, args.file))

    app = Tags(root)
    print(app.link(path, args.tag))

File: logic/tags.js
(function() {
  var ArgParse, args, fs, parser, path, print, root, src, tag;

  fs = require("graceful-fs");

  path = require("path");

  print = function(m) {
    return console.dir(m);
  };

  tag = function(root, file, tag, callback) {
    var tagDir;
    tagDir = path.join(root, tag);
    return fs.mkdir(tagDir, function(err) {
      var fileDir, fileMeta;
      if (err && !err.code === "EEXIST") {
        return callback(err, null);
      } else {
        fileMeta = path.parse(file);
        fileDir = path.join(tagDir, fileMeta.base);
        return fs.unlink(fileDir, function(err) {
          if (err && err.code === !"ENOENT") {
            return callback(err, null);
          } else {
            return fs.link(file, fileDir, function(err) {
              if (err && err.code === !"EEXIST") {
                return callback(err, null);
              } else {
                return callback(null, fileDir);
              }
            });
          }
        });
      }
    });
  };

  ArgParse = require("argparse/lib/argparse").ArgumentParser;

  parser = new ArgParse({
    version: "0.0.1",
    addHelp: true,
    description: "Tag file, moving it into folder and linking original file"
  });

  parser.addArgument(["folder"], {
    help: "The folder tags reside in."
  });

  parser.addArgument(["source"], {
    help: "The file to be tagged."
  });

  parser.addArgument(["tag"], {
    help: "The tag name"
  });

  args = parser.parseArgs();

  root = path.resolve(args.folder);

  src = path.resolve(args.source);

  tag(root, src, args.tag, function(err, path) {
    print(err);
    return print(path);
  });

}).call(this);
